### ISSUE TITLE
Alternator: performance idea: cache parsed expressions 

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#5023]
This idea was originally raised by @slivne

The DynamoDB API does not have a "prepared statement" feature like CQL's, where a request is parsed once, an ID returned and the same request can be done again with different parameters. Every DynamoDB request is one long JSON object which contains both the request, it's parameters, and everything together in one big JSON.

Nevertheless, while we don't have a choice but parse this big JSON every time, perhaps we can cache some of the parsing work inside it. For example, in an `ItemUpdate` there is an `UpdateExpression` parameter which we parse using ANTLR3, and doesn't contain the specific values given separately in an `ExpressionAttributeValues`. Perhaps we can gain performance by caching the result of this ANTLR3 parsing so next time we see exactly the same `UpdateExpression`, we don't need to call ANTLR3 again.

**Comment by nyh** [comment]
Note that our implementation today parses the expression, and then "resolves" the resulting expression (see `resolve_*()` in alternator/expressions.cc) , changing references to attribute names and constants included in the request by the actual names and constants and modifying the expression.  We need to cache the unresolved expression, not the resolved one. This adds copying of the expression, but let's make sure it's only copied once and not more than once when adding it to the cache.

**Comment by DoronArazii** [comment]
@nuivall where is it standing?

**Comment by nuivall** [comment]
The patch is in review

**Comment by nuivall** [comment]
I don't see it going to 5.4, second patch isn't even in review yet.

**Comment by avikivity** [comment]
@nyh please review it

**Comment by nuivall** [comment]
To clarify, it's not in review as I haven't yet pushed it :), as I switched to dtest effort some time ago.

**Comment by nyh** [comment]
The abandoned pull request by @nuivall, and its review comments, are in https://github.com/scylladb/scylladb/pull/14773

**Comment by mykaul** [comment]
Setting this to 6.1, knowing it might slip - we'll re-evaluate in 6.1 timeframe.

**Comment by avikivity** [comment]
Another idea: use a json parser that provides location of elements. Isolate the positions of the values (without types or quotes). Create a regular expression where each isolated value is a capture group. Then use https://github.com/VectorCamp/vectorscan (in Chimera mode) to simultaneously match all the regular expressions and extract the capture groups. We can prepare statements based on the initial parse.

To avoid compiling regular expressions prematurely, we can start by generating a hash of the shape of the query, and only compile the regex if the hash gets sufficient hits.

**Comment by nyh** [comment]
@avikivity I want to verify my understanding of your "another idea":

Is your "another idea" about how to cache the results of **JSON parsing** together with the results of the non-JSON expressions in it, and not just the expressions as suggested above?

Let's look at a concrete example, an `UpdateItem` request that I captured out of one of the tests:

```
 {
    "TableName": "alternator_Test_1733403516337",
    "Key": {"p": {"S": "2EE8RIGMHP"}},
    "UpdateExpression": "SET a = :val",
    "ExpressionAttributeValues": {":val": {"S": "hey"}}
}
```

Let's assume the table name will be part of the prepared statement (and save a few nanoseconds on not having to look up the table), so the parameters here are the key "2EE8RIGMHP" and the value is "hey". Basically after the first parsing we learn the regular expression:
```
{
    "TableName": "alternator_Test_1733403516337",
    "Key": {"p": {"S": "(.*)"}},
    "UpdateExpression": "SET a = :val",
    "ExpressionAttributeValues": {":val": {"S": "(.*)"}}
}
```

To be honest, I think there is little saving in this technique beyond the caching of just the UpdateExpression parsing (which is also trivial, but supposedly ANTLR3 is so inefficient it's worth it (?)). Note how small the JSON's structure is, will the JSON parsing really be much slower than regular expression parsing?

There is only case I can think of where the JSON can be very complex - it is when the _value_ (the thing inside the ExpressionAttributeValues) is a very complex JSON - for example, some sort of deep nested map. HOWEVER, in that case, you'd still need to parse this complex JSON value **after** finding the regular expression match, so we gain nothing.

**Comment by avikivity** [comment]
> [@avikivity](https://github.com/avikivity) I want to verify my understanding of your "another idea":
> 
> Is your "another idea" about how to cache the results of **JSON parsing** together with the results of the non-JSON expressions in it, and not just the expressions as suggested above?
> 
> Let's look at a concrete example, an `UpdateItem` request that I captured out of one of the tests:
> 
> ```
>  {
>     "TableName": "alternator_Test_1733403516337",
>     "Key": {"p": {"S": "2EE8RIGMHP"}},
>     "UpdateExpression": "SET a = :val",
>     "ExpressionAttributeValues": {":val": {"S": "hey"}}
> }
> ```
> 
> Let's assume the table name will be part of the prepared statement (and save a few nanoseconds on not having to look up the table), so the parameters here are the key "2EE8RIGMHP" and the value is "hey". Basically after the first parsing we learn the regular expression:
> 
> ```
> {
>     "TableName": "alternator_Test_1733403516337",
>     "Key": {"p": {"S": "(.*)"}},
>     "UpdateExpression": "SET a = :val",
>     "ExpressionAttributeValues": {":val": {"S": "(.*)"}}
> }
> ```
> 


Yes.

> To be honest, I think there is little saving in this technique beyond the caching of just the UpdateExpression parsing (which is also trivial, but supposedly ANTLR3 is so inefficient it's worth it (?)). Note how small the JSON's structure is, will the JSON parsing really be much slower than regular expression parsing?


It's not just the parsing - one has to construct an object out of the table name, look it up into a schema_ptr, consturct an object "UpdateExpression" and match it in some if statement, hash "SET a = :val", etc.

With regular expressions, we get a full prepared statement out of it.

> 
> There is only case I can think of where the JSON can be very complex - it is when the _value_ (the thing inside the ExpressionAttributeValues) is a very complex JSON - for example, some sort of deep nested map. HOWEVER, in that case, you'd still need to parse this complex JSON value **after** finding the regular expression match, so we gain nothing.

Do we know how much time is spent on json parsing and its immediate consequences?

Regular expression matching operates without any allocations.

**Comment by swasik** [comment]
> The abandoned pull request by [@nuivall](https://github.com/nuivall), and its review comments, are in [#14773](https://github.com/scylladb/scylladb/pull/14773)

@nyh, @nuivall why was it abandoned? Just because of Marcin being out of capacity? Or was there something wrong with assumptions for the implementation? Unfortunately I cannot find the answer in comments history.

**Comment by nyh** [comment]
> > The abandoned pull request by [@nuivall](https://github.com/nuivall), and its review comments, are in [#14773](https://github.com/scylladb/scylladb/pull/14773)
> 
> [@nyh](https://github.com/nyh), [@nuivall](https://github.com/nuivall) why was it abandoned? Just because of Marcin being out of capacity? Or was there something wrong with assumptions for the implementation? Unfortunately I cannot find the answer in comments history.

I don't know. Looking at that PR, I see that I had a complaint that I thought "loading_cache" was an overkill and made everything complicated, and also a few smaller comments - but a month later @nuivall closed the PR without explanation. We shouldn't be closing PRs just because we don't have time to finish them - open PRs don't cost us money and we can always come back to a PR a year later and push it. If there is a reason why this PR should be abandoned (e.g., Avi's idea of doing something else - not expression caching) we should say this before closing a PR silently.

**Comment by radoslawcybulski** [comment]
> {
>     "TableName": "alternator_Test_1733403516337",
>     "Key": {"p": {"S": "(.*)"}},
>     "UpdateExpression": "SET a = :val",
>     "ExpressionAttributeValues": {":val": {"S": "(.*)"}}
> }

I see some possible issues here.
1. if the regex is greedy, the regex expression might get complicated for string values (ie. "(.*)" -> "([^\"]*)" to match non string value, but for string we would have to go with "(([\].|[^\"])*)", which while not complicated will incur some performance penalty in regex. The regex will only work, if the regex engine goes left-to-right on alternatives in greedy fashion.
2. if the regex not greedy, then i'd not trust it's performance much, once we get a lot of "(.*)" in the json. I might be wrong here.
3. do we have any data on how scylla is used? Like how many different "templates" of requests are being sent? I'd imagine not many, because generating cql / dynamodb requests, which differ in format (not just values) is not that useful. Maybe it's better to assume all requests repeat (with different values)?
4. in such case manually parsing request and building some sort of trie structure from parsed data to get previously parsed request could be better?

**Comment by ScyllaPiotr** [comment]
@m-szymon implemented a simple LRU cache using `std::unordered_map` and measured its performance in the 100% cache hit scenario in a single-node cluster run on his laptop. The full results have been gathered [here](https://docs.google.com/document/d/1wm6QZsj-UZGHUbM__P4KvYdqzRGaCArS/edit#heading=h.q12mcyf816ak).

TL;DR: The current results do not justify adding a new expression cache mechanism to Alternator, as the 100% cache hit traffic resulted in ca. 5% reduction in total processing time of a request.

**Comment by avikivity** [comment]
5% is quite a large win for a simple cache.

Consider 10 5% wins, you won't make any of them because you think they're not justified, so you lose a 50% win.

In any case, I think caching the entire request is better.

**Comment by ScyllaPiotr** [comment]
This 5% is when the cache has 100% hit ratio and a single element on it.
And the cluster is composed of just a single node.
The cache will be a new mechanism, with such a small margin it will require careful tuning to actually improve performance and will add to complexity. It will require separate tests.

Regarding caching of the entire request, I'll let @m-szymon respond (after 21 July).

**Comment by avikivity** [comment]
> This 5% is when the cache has 100% hit ratio and a single element on it.

100% hit ratio is expected. The cache will only miss when the application changes or when a node starts.

> And the cluster is composed of just a single node. The cache will be a new mechanism, with such a small margin it will require careful tuning to actually improve performance and will add to complexity. It will require separate tests.

5% is not a small margin (though I believe much more can be achieved).

I don't see why any careful tuning needed.


> 
> Regarding caching of the entire request, I'll let [@m-szymon](https://github.com/m-szymon) respond (after 21 July).

Ok, I'll respond to his response on 42 August

**Comment by nuivall** [comment]
> 5% is not a small margin (though I believe much more can be achieved).

This scales with expression complexity. E.g. for UpdateExpression with 10 fields I had 15% improvement. And usage of parenthesis may add even more I believe.

**Comment by nyh** [comment]
> This scales with expression complexity. E.g. for UpdateExpression with 10 fields I had 15% improvement. And usage of parenthesis may add even more I believe.

How did you measure this 15% improvement? Was it on a single-node cluster or a larger cluster?
Note that UpdateItem is a fairly expensive operation on a real cluster - involving a read before write done by LWT (so it's CL=ALL, right?) the whole Paxos schtick for LWT, and a write on all RF replicas. All are very expensive for clusters that don't have a single node. If you measured the 5% or 15% improvement on a single-node cluster, the real improvement might be significantly lower on a real cluster.

**Comment by nuivall** [comment]
>  Each test repeated 10x on a fresh scylla instance (single node)

As I understand the doc mentioned 5% is also on a single node.

**Comment by m-szymon** [comment]
> The current results do not justify adding a new expression cache mechanism to Alternator,

Actually my conclusions were a little bit different - with the most basic expression caching it should give some small gain and it is quite simple, so it is worth to add it, but overall (in general case) this part of processing is not that time consuming (relative to entire request), that further optimizations may not be worth it (now).

But the actual numbers very much depend on what we consider our reference. On one hand with some complex expressions we may measure higher values. But what we measured is close to highest possible value for given request - not achievable on real cluster, as @nyh wrote.

**Comment by swasik** [comment]
> But the actual numbers very much depend on what we consider our reference. On one hand with some complex expressions we may measure higher values. But what we measured is close to highest possible value for given request - not achievable on real cluster, as [@nyh](https://github.com/nyh) wrote.

So could we measure it on the real cluster?

**Comment by avikivity** [comment]
> > But the actual numbers very much depend on what we consider our reference. On one hand with some complex expressions we may measure higher values. But what we measured is close to highest possible value for given request - not achievable on real cluster, as [@nyh](https://github.com/nyh) wrote.
> 
> So could we measure it on the real cluster?

perf-alternator

**Comment by mykaul** [comment]
With https://github.com/scylladb/scylladb/commit/1aef733d481caf7978ffb3c65c508378c8049484 merged, what else is needed to complete this?

**Comment by ScyllaPiotr** [comment]
Per https://github.com/scylladb/scylladb/issues/5023#issuecomment-3052680942 and with the elaborate discussion here,
let's rename the issue into "Alternator: performance idea: cache requests"
and close it with a future, already discussed PR that will add caching of the whole requests.
Nadav has already thought about a subsequent step (after this renamed issue is done): https://github.com/scylladb/scylladb/issues/26015

**Comment by mykaul** [comment]
@ScyllaPiotr - so the milestone 2025.4 is no longer relevant here?

**Comment by ScyllaPiotr** [comment]
Yes, should you agree with my proposal, the milestone should be changed into 2026.1.

**Comment by swasik** [comment]
> let's rename the issue into "Alternator: performance idea: cache requests" and close it with a future, already discussed PR that will add caching of the whole requests. 

@ScyllaPiotr I am not sure I understand. You want to implement one more PR to close this issue?

IMO, as we already merged partial solution for this, let us close this Issue and create the new one where you will clearly describe the scope of remaining changes (assigned to 2026.1). This discussion here starts to be difficult to follow.

**Comment by ScyllaPiotr** [comment]
Following #25176 merged, closing this one.
Follow-up issue: https://github.com/scylladb/scylladb/issues/26703

**Comment by nyh** [comment]
My three cents:
1. It was a mistake for the pull request to Refs, instead of Fixes, this issue. I'm surprised I didn't notice this earlier, I usually complain when I see a "Refs" instead of "Fixes" with no explanation of why we want the issue to remain open. This issue was always about caching parsed expressions - nothing else - and this pull request fixed it. So this issue should have been closed.
2. Avi had another idea to use regular expressions and pieces of the entire JSON - not just expressions. This is a different idea, and if Avi really believes in it he or someone else should create a new issue for that idea - it shouldn't keep this issue open. You can create the new issue based on stuff that Avi wrote in this issue.
3. Amnon had yet another idea to add support for a prepared-statement-like feature that modified drivers will use, and will make requests shorter and faster to parse. I created an issue based on my understanding of his idea: https://github.com/scylladb/scylladb/issues/26015
